File: kotlin/src/main/kotlin/TennisGame2.kt
import java.lang.RuntimeException

class TennisGame2(private val player1Name: String, private val player2Name: String) : TennisGame {

    var P1point: Int = 0
    var P2point: Int = 0

    override fun getScore(): String {
        if (P1point.hasWon(P2point)) {
            return "Win for player1"
        }
        if (P2point.hasWon(P1point)) {
            return "Win for player2"
        }

        if (P1point.isDeuce(P2point))
            return "Deuce"

        if (P1point.hasAdvantage(P2point))
            return "Advantage player1"

        if (P2point.hasAdvantage(P1point))
            return "Advantage player2"

        if (P1point < 4 && P2point < 4) {
            val score1 = P1point.getScoreAsString()
            val score2 = P2point.getScoreAsString()
            return if (score1 == score2)
                "$score1-All"
            else
                "$score1-$score2"
        }

        throw RuntimeException("Invalid game result")
    }

    private fun Int.hasAdvantage(p2point: Int) = this > p2point && p2point >= 3
    private fun Int.isDeuce(p2point: Int) = this == p2point && this >= 3

    private fun Int.hasWon(competitor: Int) = this >= 4 && competitor >= 0 && this - competitor >= 2

    private fun Int.getScoreAsString(): String? = when (this) {
        0 -> "Love"
        1 -> "Fifteen"
        2 -> "Thirty"
        3 -> "Forty"
        else -> null
    }

    fun SetP1Score(number: Int) {

        for (i in 0 until number) {
            P1Score()
        }

    }

    fun SetP2Score(number: Int) {

        for (i in 0 until number) {
            P2Score()
        }

    }

    fun P1Score() {
        P1point++
    }

    fun P2Score() {
        P2point++
    }

    override fun wonPoint(player: String) {
        if (player === "player1")
            P1Score()
        else
            P2Score()
    }
}

File: php/src/TennisGame2.php
<?php

declare(strict_types=1);

namespace TennisGame;

class TennisGame2 implements TennisGame
{
    private int $P1point = 0;

    private int $P2point = 0;

    private string $P1res = '';

    private string $P2res = '';

    public function __construct(
        private string $player1Name,
        private string $player2Name
    ) {
    }

    public function getScore(): string
    {
        $score = '';
        if ($this->P1point === $this->P2point && $this->P1point < 4) {
            if ($this->P1point === 0) {
                $score = 'Love';
            }
            if ($this->P1point === 1) {
                $score = 'Fifteen';
            }
            if ($this->P1point === 2) {
                $score = 'Thirty';
            }
            $score .= '-All';
        }

        if ($this->P1point === $this->P2point && $this->P1point >= 3) {
            $score = 'Deuce';
        }

        if ($this->P1point > 0 && $this->P2point === 0) {
            if ($this->P1point === 1) {
                $this->P1res = 'Fifteen';
            }
            if ($this->P1point === 2) {
                $this->P1res = 'Thirty';
            }
            if ($this->P1point === 3) {
                $this->P1res = 'Forty';
            }

            $this->P2res = 'Love';
            $score = "{$this->P1res}-{$this->P2res}";
        }

        if ($this->P2point > 0 && $this->P1point === 0) {
            if ($this->P2point === 1) {
                $this->P2res = 'Fifteen';
            }
            if ($this->P2point === 2) {
                $this->P2res = 'Thirty';
            }
            if ($this->P2point === 3) {
                $this->P2res = 'Forty';
            }
            $this->P1res = 'Love';
            $score = "{$this->P1res}-{$this->P2res}";
        }

        if ($this->P1point > $this->P2point && $this->P1point < 4) {
            if ($this->P1point === 2) {
                $this->P1res = 'Thirty';
            }
            if ($this->P1point === 3) {
                $this->P1res = 'Forty';
            }
            if ($this->P2point === 1) {
                $this->P2res = 'Fifteen';
            }
            if ($this->P2point === 2) {
                $this->P2res = 'Thirty';
            }
            $score = "{$this->P1res}-{$this->P2res}";
        }

        if ($this->P2point > $this->P1point && $this->P2point < 4) {
            if ($this->P2point === 2) {
                $this->P2res = 'Thirty';
            }
            if ($this->P2point === 3) {
                $this->P2res = 'Forty';
            }
            if ($this->P1point === 1) {
                $this->P1res = 'Fifteen';
            }
            if ($this->P1point === 2) {
                $this->P1res = 'Thirty';
            }
            $score = "{$this->P1res}-{$this->P2res}";
        }

        if ($this->P1point > $this->P2point && $this->P2point >= 3) {
            $score = 'Advantage player1';
        }

        if ($this->P2point > $this->P1point && $this->P1point >= 3) {
            $score = 'Advantage player2';
        }

        if ($this->P1point >= 4 && $this->P2point >= 0 && ($this->P1point - $this->P2point) >= 2) {
            $score = 'Win for player1';
        }

        if ($this->P2point >= 4 && $this->P1point >= 0 && ($this->P2point - $this->P1point) >= 2) {
            $score = 'Win for player2';
        }

        return $score;
    }

    public function wonPoint(string $player): void
    {
        if ($player === 'player1') {
            $this->P1Score();
        } else {
            $this->P2Score();
        }
    }

    private function SetP1Score(int $number): void
    {
        for ($i = 0; $i < $number; $i++) {
            $this->P1Score();
        }
    }

    private function SetP2Score(int $number): void
    {
        for ($i = 0; $i < $number; $i++) {
            $this->P2Score();
        }
    }

    private function P1Score(): void
    {
        $this->P1point++;
    }

    private function P2Score(): void
    {
        $this->P2point++;
    }
}
